src/lib/FreeC/Pass/SharingAnalysisPass.hs

```diff
@@ -0,0 +1,144 @@
+analyseSharingDecl :: IR.FuncDecl -> Converter IR.FuncDecl
+analyseSharingDecl (IR.FuncDecl srcSpan ident typeArgs args returnType rhs) = do
+    let varList = (map (\p -> ( fst p, snd $ snd p )) . filter
+                   ((> 1) . fst . snd) . Map.toList . countVarNames) rhs
+    rhs' <- buildLet rhs varList
+    return (IR.FuncDecl srcSpan ident typeArgs args returnType rhs')
+
+-- | Builds a @let@-expression from the given expression and variable names.
+--
+--   Computes @let@-bindings from the given variables, composes the resulting
+--   substitutions and applies the substitution on the expression.
+buildLet
+    :: IR.Expr -> [ ( IR.VarName, Maybe IR.TypeScheme ) ] -> Converter IR.Expr
+buildLet expr []   = return expr
+buildLet expr vars = do
+    let srcSpan = IR.exprSrcSpan expr
+    ( binds, substs ) <- buildBinds srcSpan vars
+    return (IR.Let srcSpan binds (applySubst (composeSubsts substs) expr)
+            (IR.exprTypeScheme expr))
+
+-- | Converts the list containing variables into @let@-bindings where
+--   the variable pattern is a fresh variable and the right-hand side is
+--   a variable that occurred multiple times on right hand sides.
+--
+--   Also computes substitutions mapping given variables to fresh variables.
+--   Returns the generated let-bindings and the substitutions.
+buildBinds :: SrcSpan -> [ ( IR.VarName, Maybe IR.TypeScheme ) ]
+    -> Converter ( [ IR.Bind ], [ Subst IR.Expr ] )
+buildBinds srcSpan = foldM buildBind ( [], [] )
+  where
+    buildBind :: ( [ IR.Bind ], [ Subst IR.Expr ] )
+        -> ( IR.VarName, Maybe IR.TypeScheme )
+        -> Converter ( [ IR.Bind ], [ Subst IR.Expr ] )
+    buildBind ( binds, substs ) ( varName, varTypeScheme ) = do
+        varName' <- freshHaskellQName varName
+        let subst = singleSubst' varName (\s -> IR.Var s varName' varTypeScheme)
+            -- This type scheme has wrong srcSpanInformation
+            rhs = IR.Var srcSpan varName varTypeScheme
+            Just varPatIdent = IR.identFromQName varName'
+            varPatType = fmap IR.typeSchemeType varTypeScheme
+            varPat = IR.VarPat srcSpan varPatIdent varPatType False
+            bind = IR.Bind srcSpan varPat rhs
+        _ <- renameAndDefineVar srcSpan False varPatIdent varPatType
+        return ( bind : binds, subst : substs )
+
+-- | Counts all variable names on right-hand sides of expression.
+--
+--   If a variable is annotated with a type scheme the annotation is introduced
+--   into the map as well.
+countVarNames :: IR.Expr -> Map IR.VarName ( Integer, Maybe IR.TypeScheme )
+countVarNames IR.Con {} = Map.empty
+countVarNames (IR.Var _ varName maybeTypeScheme)
+    = Map.singleton varName ( 1, maybeTypeScheme )
+countVarNames (IR.App _ lhs rhs _)
+    = countVarNames lhs `mergeMap` countVarNames rhs
+countVarNames (IR.TypeAppExpr _ lhs _ _) = countVarNames lhs
+countVarNames (IR.If _ e1 e2 e3 _) = (countVarNames e1)
+    `mergeMap` (countVarNames e2) `mergeMap` countVarNames e3
```

Comment:
Looks like the CI pipeline has not been triggered for some reason. HLint should have complained about the redundant parentheses.